File: src/160.intersection-of-two-linked-lists/solution_hash-table.ts
/*
 * @lc app=leetcode.cn id=160 lang=typescript
 *
 * [160] 相交链表
 */

interface ListNode {
  val: number;
  next: ListNode | null;
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function getIntersectionNode(
  l1: ListNode | null,
  l2: ListNode | null,
): ListNode | null {
  if (!l1 || !l2) {
    return null;
  }

  const set = new Set<ListNode>();
  let p1: ListNode | null = l1;
  let p2: ListNode | null = l2;

  while (p1 || p2) {
    if (p1) {
      if (set.has(p1)) {
        return p1;
      } else {
        set.add(p1);
        p1 = p1.next;
      }
    }

    if (p2) {
      if (set.has(p2)) {
        return p2;
      } else {
        set.add(p2);
        p2 = p2.next;
      }
    }
  }

  return null;
}
// @lc code=end

export default getIntersectionNode;

File: src/138.copy-list-with-random-pointer/solution_hash-table.cpp
/*
 * @lc app=leetcode.cn id=138 lang=cpp
 *
 * https://leetcode-cn.com/problems/copy-list-with-random-pointer/
 */

#include <unordered_map>

using namespace std;

/**
 * Definition for a Node.
 */
class Node {
 public:
  int val;
  Node *next;
  Node *random;

  Node(int _val) {
    val = _val;
    next = nullptr;
    random = nullptr;
  }
};

// @lc code=start
class Solution {
 public:
  Node *copyRandomList(Node *head) {
    if (!head) {
      return nullptr;
    }

    unordered_map<Node *, Node *> dict = {{head, new Node(head->val)}};
    Node *curr = head;

    while (curr) {
      Node *node = dict.find(curr)->second;

      if (curr->random) {
        if (dict.find(curr->random) == dict.end()) {
          dict[curr->random] = new Node(curr->random->val);
        }
        node->random = dict[curr->random];
      }

      if (curr->next) {
        if (dict.find(curr->next) == dict.end()) {
          dict[curr->next] = new Node(curr->next->val);
        }
        node->next = dict[curr->next];
      }

      curr = curr->next;
    }

    return dict.find(head)->second;
  }
};
// @lc code=end
